File: sponsio/render/tokens.py
"""Single source of truth for colors, symbols, and styles.

Every other module in ``sponsio/render/`` imports from here. Adding a
new color or symbol means adding a token here with a usage rationale —
not sprinkling raw hex codes through the codebase.

The palette and service colors are taken from the v1 CLI redesign
spec; the values are also referenced by tests that enforce the "no
raw hex outside tokens.py" invariant.
"""

from __future__ import annotations

# ---------------------------------------------------------------------------
# Semantic palette — what each color *means*, not what it is.
# ---------------------------------------------------------------------------

PALETTE: dict[str, str] = {
    "brand": "#7DD3FC",  # cyan-300   — banner titles, CTA arrow
    "success": "#86EFAC",  # green-300  — ✓, ACTIVE, PASS
    "violation": "#FCA5A5",  # red-300    — ✗, BLOCKED, VIOLATED
    "warning": "#FCD34D",  # amber-300  — soft fail, watchlist
    "active": "#C4B5FD",  # violet-300 — contract state transitions
    "metadata": "#64748B",  # slate-500  — timestamps, latencies, hints
    "muted": "#94A3B8",  # slate-400  — fallback service label, dim text
    "rule": "#475569",  # slate-600  — banner & divider lines
    "fg": "#E2E8F0",  # slate-200  — default foreground
}

# ---------------------------------------------------------------------------
# Service brand colors — capped to ~30% saturation so labels never
# overpower verdicts. Unknown services fall back to ``muted``.
# ---------------------------------------------------------------------------

# Transport color palette — the only labels ``service_for_tool`` ever
# returns. ``function`` is the background-noise default (muted);
# distinguished transports get a distinct hue so they pop in the trace.
SERVICE_COLORS: dict[str, str] = {
    "func": PALETTE["muted"],
    "shell": "#F57C00",  # orange — agent shelling out is high-risk
    "mcp": "#7B1FA2",  # magenta
    "http": "#1976D2",  # blue
}


def service_color(name: str) -> str:
    """Return the brand color for ``name``, or the muted fallback."""
    return SERVICE_COLORS.get(name, PALETTE["muted"])


# ---------------------------------------------------------------------------
# Symbol vocabulary — see spec §3.1. Use ONLY these.
# ---------------------------------------------------------------------------

SYMBOLS: dict[str, str] = {
    "pass": "✓",
    "fail": "✗",
    "active": "⚙",
    "cta": "→",
    "tree_branch": "├─",
    "tree_continue": "│",
    "tree_end": "└─",
    "rule_heavy": "━",
    "rule_light": "─",
    # Brand mark — the half-filled circle pair from the original
    # `print_banner`. Sits between the heavy rule and the brand name.
    "logo": "◒◓",
}


# ---------------------------------------------------------------------------
# Verdict status words — see spec §3.2. Always uppercase, always one of
# these. Add a new word here before using it.
# ---------------------------------------------------------------------------

STATUS = {
    "READY": PALETTE["success"],
    "ACTIVE": PALETTE["success"],
    "PASS": PALETTE["success"],
    "BLOCKED": PALETTE["violation"],
    "WARN": PALETTE["warning"],
    "EXPIRED": PALETTE["metadata"],
}


# ---------------------------------------------------------------------------
# SVG export theme — applied when Console.save_svg() is called.
# ---------------------------------------------------------------------------

SVG_THEME = {
    "background": "#0F172A",  # slate-900
    "foreground": PALETTE["fg"],
}

File: ts/packages/sdk/src/cli/tokens.ts
/**
 * Truecolor palette mirroring ``sponsio/render/tokens.py`` so the TS
 * scanner's stdout matches the Python wizard / runtime trace output
 * pixel-for-pixel.  Single source of truth on the TS side — every
 * other module in this package imports these helpers instead of
 * sprinkling raw ``\x1b[31m`` / ``\x1b[32m`` codes.
 *
 * If you tweak the Python PALETTE, mirror the change here.  The
 * cross-language test suite doesn't enforce parity (no shared
 * runner), so it's a discipline thing.
 */

/* ─────────────────────────────────────────────────────────────────
 * Truecolor escape helpers — keep payload structure identical to
 * Python's so palette diffs read cleanly.
 * ──────────────────────────────────────────────────────────────── */

const RESET = "\x1b[0m";

function rgb(hex: string): string {
  // ``#7DD3FC`` → ``\x1b[38;2;125;211;252m``
  const m = /^#?([0-9a-f]{2})([0-9a-f]{2})([0-9a-f]{2})$/i.exec(hex);
  if (!m) throw new Error(`bad hex: ${hex}`);
  const [, r, g, b] = m;
  return `\x1b[38;2;${parseInt(r, 16)};${parseInt(g, 16)};${parseInt(b, 16)}m`;
}

/* ─────────────────────────────────────────────────────────────────
 * Palette — keep 1:1 with sponsio/render/tokens.py:PALETTE.
 * ──────────────────────────────────────────────────────────────── */

export const PALETTE = {
  brand: rgb("#7DD3FC"), // cyan-300   — banner titles, CTA arrow
  success: rgb("#86EFAC"), // green-300  — ✓, ACTIVE, PASS
  violation: rgb("#FCA5A5"), // red-300    — ✗, BLOCKED, VIOLATED
  warning: rgb("#FCD34D"), // amber-300  — soft fail, watchlist
  active: rgb("#C4B5FD"), // violet-300 — contract state transitions
  metadata: rgb("#64748B"), // slate-500  — timestamps, latencies, hints
  muted: rgb("#94A3B8"), // slate-400  — fallback service label, dim text
  rule: rgb("#475569"), // slate-600  — banner & divider lines
  fg: rgb("#E2E8F0"), // slate-200  — default foreground
} as const;

const BOLD = "\x1b[1m";

/* ─────────────────────────────────────────────────────────────────
 * Convenience wrappers — every place that wants to colorize a string
 * goes through these.  Centralized so we can short-circuit when
 * stdout isn't a TTY (`NO_COLOR` env var, piped output) without
 * touching every call site.
 * ──────────────────────────────────────────────────────────────── */

const colorEnabled = (): boolean => {
  if (process.env.NO_COLOR) return false;
  // ``TERM=dumb`` is the universal "no escapes" hint.
  if (process.env.TERM === "dumb") return false;
  // When stdout is a pipe (`| cat`, CI logs) most colour systems
  // turn off; do the same so byte-identical comparisons in tests
  // don't have to strip escapes.
  return Boolean(process.stdout.isTTY);
};

function paint(color: string, text: string, { bold = false }: { bold?: boolean } = {}): string {
  if (!colorEnabled()) return text;
  return (bold ? BOLD : "") + color + text + RESET;
}

export const style = {
  brand: (s: string, bold = false) => paint(PALETTE.brand, s, { bold }),
  success: (s: string, bold = false) => paint(PALETTE.success, s, { bold }),
  violation: (s: string, bold = false) => paint(PALETTE.violation, s, { bold }),
  warning: (s: string, bold = false) => paint(PALETTE.warning, s, { bold }),
  metadata: (s: string) => paint(PALETTE.metadata, s),
  fg: (s: string) => paint(PALETTE.fg, s),
};
